src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,416 @@
 ---
-title: Null safety migration guide
-description: Watch this space for information on how to make your code null safe
-toc: false
+title: Migrating to null safety
+description: How to move your existing Dart code to the world of null safety
 ---
 
-When the migration tool for null safety is available,
-this page will have information on how to
-change your production code to be null safe.
-Until then, you're welcome to try out null safety,
-but please don't depend on it.
+{{ site.alert.warning }}
+**This page is a DRAFT.**
+Please give us feedback at
+[PR 2739](https://github.com/dart-lang/site-www/pull/2739).
+{{ site.alert.end }}
 
-{{site.alert.important}}
-  Watch the [Dart announcements][] group and the [Dart blog][]
-  for announcements about null safety.
-{{site.alert.end}}
+This page describes how and when to migrate your code to [null safety][].
+Here are the basic steps for migrating each package that you own:
 
-For more information, see the [null safety homepage][].
+1. **Wait** for the packages
+   that you depend on to migrate.
+2. **Migrate** your package's code,
+   preferably using the interactive migration tool (`dart migrate`).
+3. **Test** to make sure your changes work.
+4. If the package is already on pub.dev,
+   **publish** the null-safe version as a **prerelease** version.
 
-[null safety homepage]: /null-safety
-[Dart announcements]: {{site.group}}/d/forum/announce
-[Dart blog]: https://medium.com/dartlang
+{{ site.alert.info }}
+  **Migrating an app is the same as migrating a package.**
+  Before fully migrating an app,
+  consider waiting until null safety is in a stable release.
+  In the meantime, you can add migration hints to your app
```

Comment:
Do we expect them to add migration hints manually?  The tool will try to upgrade them (prematurely in this case), but it also presents a handy way to see where hints might be needed or useful...

@stereotype441 we may want to support a mode in which users can fire up the migration tool for the sole purpose of seeing what would happen, and adding hints.  This can be done as-is by simply not applying the suggestions, but is that an intended/well-supported use case?

@kwalrath if we expect them to do so manually, then making "migration hints" link to a page or section that documents what the supported hints are would be useful.

Reply:
I'm not sure it's worthwhile to generally recommend several migration passes to most app developers. I'd rather loosen the guidance against migrating apps.

Reply:
@kwalrath to clarify, I'm suggesting we change:

"Migrating an app is the same as migrating a package. Before fully migrating an app, consider waiting until null safety is in a stable release. In the meantime, you can learn about null safety, perhaps adding migration hints to your app or refactoring your app code."

to:

"Migrating an app is technically the same as migrating a package. Before fully migrating an app, consider waiting until null safety is in a stable release, and all your dependencies are ready. In the meantime, you can learn about null safety."